### PR TITLE
Script for Network virtualized devices

### PR DESCRIPTION
**Pull request by hathyaga**
This script will add a Network virtualized device
then does an add and remove of a Backing device
and finally removes the Network virtualized device
This will be performed as many times based on user
input

Signed-off-by: Harsha Thyagaraja <harshkid@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by hathyaga** [comment]
Job log 
[job.log.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/1863606/job.log.txt)

**Comment by hathyaga** [comment]
@narasimhan-v  This PR has Backing device add/remove included to Network_virtualization_dlpar.py script. 

Please review.

File: io/net/virt-net/Network_virtualization_dlpar.py [hunk 1]
```diff
@@ -160,34 +170,49 @@ def test(self):
             self.fail("Failed to start the rsct and rsct_rm services")
 
         for _ in range(self.count):
-            self.network_virtualization__add()
+            self.network_virtualization_add()
+            self.network_virtualization_backing_device_add()
+            self.network_virtualization_backing_device_remove()
```
Comment (narasimhan-v) [inline io/net/virt-net/Network_virtualization_dlpar.py]: Ideally, these 4 should be 4 seperate `test_x()` functions

File: io/net/virt-net/Network_virtualization_dlpar.py [hunk 2]
```diff
@@ -198,8 +223,9 @@ def listhwres(self, server, lpar, slot_num):
             self.fail("lshwres operation failed ")
         return output
 
-    def changehwres(self, server, operation, lpar_id, slot_num,
-                    backing_devices, msg):
+    def Network_virtualization_add_device(self, server, operation,
```
Comment (narasimhan-v) [inline io/net/virt-net/Network_virtualization_dlpar.py]: It should not have `add` in name, as it removes as well

File: io/net/virt-net/Network_virtualization_dlpar.py [hunk 3]
```diff
@@ -212,7 +238,70 @@ def changehwres(self, server, operation, lpar_id, slot_num,
...
+        '''
+        pre_add = self.backing_dev_count(self.slot_num)
+        self.backing_dev_add(self.server, 'a', self.lpar_id,
+                             self.slot_num, self.add_backing_device,
+                             'add')
+        post_add = self.backing_dev_count(self.slot_num)
+        if post_add - pre_add != 1:
+            self.fail("Failed to add backing device")
+
+    def backing_dev_add(self, server, operation, lpar_id, slot_num,
```
Comment (narasimhan-v) [inline io/net/virt-net/Network_virtualization_dlpar.py]: function name change here as well. no `add`

Comment (narasimhan-v) [reply io/net/virt-net/Network_virtualization_dlpar.py]: docstring missing

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 2 above)
Comment (narasimhan-v) [reply io/net/virt-net/Network_virtualization_dlpar.py]: docstring missing

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 1 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: When I prefix 'test' to it., I get 5 variants of tests being run. but they run based on the order in which they occur in the script. 


[root@ltc-zz3-lp15 TEST]# avocado run Network_virtualization.py -m Network_virtualization.py.data/Network_virtualization.yaml 
JOB ID     : 0f06e8a8486ad4b1036152f04085ede68707384d
JOB LOG    : /root/avocado/job-results/job-2018-03-30T08.54-0f06e8a/job.log
 (1/5) Network_virtualization.py:NetworkVirtualization.test;run-62e5: FAIL (45.74 s)
 (2/5) Network_virtualization.py:NetworkVirtualization.test_network_virtualization_remove;run-62e5: PASS (79.62 s)
 (3/5) Network_virtualization.py:NetworkVirtualization.test_network_virtualization_add;run-62e5: PASS (56.94 s)
 (4/5) Network_virtualization.py:NetworkVirtualization.test_network_virtualization_backing_device_remove;run-62e5: FAIL (37.33 s)
 (5/5) Network_virtualization.py:NetworkVirtualization.test_network_virtualization_backing_device_add;run-62e5: PASS (53.82 s)
RESULTS    : PASS 3 | ERROR 0 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 273.66 s


This is not the right order. Need suggestion on what can be done to corrcet the order of the script run

Comment (narasimhan-v) [reply io/net/virt-net/Network_virtualization_dlpar.py]: I don't think `test()` is required anymore.
You can re-order the functions. You are free to choose the order logically.
Example:
```
test_network_virtualization_add
test_network_virtualization_remove
test_network_virtualization_backing_device_add
test_network_virtualization_backing_device_remove
```

Comment (narasimhan-v) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Also, the existing `service start` part can be transformed into a function and called from `setUp()` instead.

**Comment by narasimhan-v** [comment]
@hathyaga https://github.com/hathyaga/avocado-misc-tests-1/blob/075498b03a982c695bb0fdcead2e5165eafd798f/io/net/virt-net/Network_virtualization_dlpar.py
should also be removed via this commit.
Ideally, renamed and added functionality.

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 1 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Ok. I will implement this. 

Right now, Stuck at something really simple, (I guess)

    for _ in range(self.count):
        self.test_network_virtualization_add()
        #self.test_network_virtualization_backing_device_add()
        #self.test_network_virtualization_backing_device_remove()
        #self.test_network_virtualization_remove()


I see the below error:
    for _ in range(self.count):
NameError: name 'self' is not defined


Even when I have defined count as this:
self.count = int(self.params.get("count", '*', default="1"))

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 2 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Sure. Will make those changes

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 3 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Will make those changes

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 1 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Sure. Will include these changes.

Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: @narasimhan-v After including the yaml parameter, 
iteration: !mux
    1:
    2:

I found the tests to run in this order

test1
test1

test2
test2

===> Finding out how I can run it in order
test1
test2

test1
test2

**Comment by hathyaga** [comment]
All suggested changes are addressed. Please review

**Comment by hathyaga** [comment]
New job log
[job.log.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/1868260/job.log.txt)

File: io/net/virt-net/Network_virtualization_dlpar.py (hunk 1 above)
Comment (hathyaga) [reply io/net/virt-net/Network_virtualization_dlpar.py]: Used --execution-order tests-per-variant in the command line to get this working. 
This has been  documented in README

File: io/net/virt-net/Network_virtualization.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+            for svc in ["rsct", "rsct_rm"]:
+                process.run('startsrc -g %s' % svc, shell=True, sudo=True)
+        except CmdError as details:
+            self.log.debug(str(details))
+            self.fail("Starting service %s failed", svc)
+
+        output = process.system_output("lssrc -a", ignore_status=True,
+                                       shell=True, sudo=True)
+        if "inoperative" in output:
+            self.fail("Failed to start the rsct and rsct_rm services")
```
Comment (narasimhan-v) [inline io/net/virt-net/Network_virtualization.py]: it should be `self.skip()`

File: io/net/virt-net/Network_virtualization_dlpar.py [hunk 5]
```diff
@@ -45,7 +45,8 @@ def __str__(self):
 
 class NetworkVirtualization(Test):
     '''
-    Adding and deleting Network Virtualized ports from the vios
+    Adding and deleting Network Virtualized devices from the vios
```
Comment (narasimhan-v) [inline io/net/virt-net/Network_virtualization_dlpar.py]: This script should be renamed, and changes go into network_virtualization.py

**Comment by hathyaga** [comment]
New PR with all the changes requested
https://github.com/avocado-framework-tests/avocado-misc-tests/pull/842

**Pull request closed without merging**